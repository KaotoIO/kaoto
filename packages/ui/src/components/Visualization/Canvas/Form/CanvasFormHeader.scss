.form-header {
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  margin-bottom: 15px;

  img[class^='form-header__icon-'] {
    height: 40px;
    margin-right: 10px;
  }

  &__title {
    text-transform: capitalize;
  }
}

.form-tabs {
  display: flex;
  flex-wrap: wrap;
  margin-bottom: 15px;

  /* stylelint-disable-next-line selector-class-pattern */
  .pf-v5-c-toggle-group__item {
    display: flex;
    flex: content;
    min-width: 33%;
  }

  button {
    flex: 1;
    justify-content: center;
    font-weight: bold;

    --pf-v5-c-toggle-group__button--PaddingRight: 0.25rem;
    --pf-v5-c-toggle-group__button--PaddingLeft: 0.25rem;
  }
}
