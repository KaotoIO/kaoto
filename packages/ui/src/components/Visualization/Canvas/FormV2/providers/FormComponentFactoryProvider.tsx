import { createContext, FunctionComponent, PropsWithChildren, useCallback } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';
import { BooleanField } from '../fields/BooleanField';
import { DisabledField } from '../fields/DisabledField';
import { ObjectField } from '../fields/ObjectField/ObjectField';
import { OneOfField } from '../fields/OneOfField/OneOfField';
import { StringField } from '../fields/StringField';
import { PasswordField } from '../fields/PasswordField';
import { FieldProps } from '../typings';

type FormComponentFactoryContextValue = (schema: KaotoSchemaDefinition['schema']) => FunctionComponent<FieldProps>;

export const FormComponentFactoryContext = createContext<FormComponentFactoryContextValue | undefined>(undefined);

export const FormComponentFactoryProvider: FunctionComponent<PropsWithChildren> = ({ children }) => {
  const factory = useCallback<FormComponentFactoryContextValue>((schema) => {
    if (schema.format === 'password') {
      return PasswordField;
    }
    switch (schema.type) {
      case 'string':
      case 'number':
      case 'integer':
        return StringField;
      case 'boolean':
        return BooleanField;
      case 'object':
        return ObjectField;
      case 'array':
        return () => <div>Array field</div>;
    }

    if (Array.isArray(schema.oneOf)) {
      return OneOfField;
    } else if (Array.isArray(schema.anyOf)) {
      return () => <div>AnyOf field</div>;
    }

    return DisabledField;
  }, []);

  return <FormComponentFactoryContext.Provider value={factory}>{children}</FormComponentFactoryContext.Provider>;
};
