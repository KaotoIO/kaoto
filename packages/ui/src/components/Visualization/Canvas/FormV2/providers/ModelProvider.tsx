import { createContext, FunctionComponent, PropsWithChildren, useMemo } from 'react';

export interface ModelContextValue {
  model: unknown;
  errors?: Record<string, string[]>;
  onPropertyChange: (propName: string, value: unknown) => void;
  disabled?: boolean;
}

export const ModelContext = createContext<ModelContextValue>({ model: {}, onPropertyChange: () => {} });

export const ModelContextProvider: FunctionComponent<PropsWithChildren<ModelContextValue>> = ({
  model,
  errors,
  onPropertyChange,
  disabled,
  children,
}) => {
  const value = useMemo(() => {
    return { model, errors, onPropertyChange, disabled };
  }, [model, errors, onPropertyChange, disabled]);

  return <ModelContext.Provider value={value}>{children}</ModelContext.Provider>;
};
