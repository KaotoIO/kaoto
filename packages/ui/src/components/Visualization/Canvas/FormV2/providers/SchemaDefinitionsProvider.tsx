import { createContext, FunctionComponent, PropsWithChildren } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';

type SchemaDefinitionsContextValue = Record<string, KaotoSchemaDefinition['schema']>;

export const SchemaDefinitionsContext = createContext<SchemaDefinitionsContextValue>({});

export const SchemaDefinitionsProvider: FunctionComponent<
  PropsWithChildren<{ schema: KaotoSchemaDefinition['schema'] }>
> = ({ schema, children }) => {
  const definitions: SchemaDefinitionsContextValue = schema.definitions ?? {};

  return <SchemaDefinitionsContext.Provider value={definitions}>{children}</SchemaDefinitionsContext.Provider>;
};
