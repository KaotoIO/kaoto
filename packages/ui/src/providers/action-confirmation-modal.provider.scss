.pf-v5-c-modal-box {
  &__title-text {
    white-space: normal;
  }
}
