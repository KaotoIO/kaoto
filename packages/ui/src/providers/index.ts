export * from './catalog-schema-loader.provider';
export * from './entities.provider';
