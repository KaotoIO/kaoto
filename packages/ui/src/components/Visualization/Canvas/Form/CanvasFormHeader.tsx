import { Button, Grid, GridItem, SearchInput, Title, ToggleGroup, ToggleGroupItem } from '@patternfly/react-core';
import { TimesIcon } from '@patternfly/react-icons';
import { FunctionComponent, useContext } from 'react';
import { CanvasFormTabsContext } from '../../../../providers/canvas-form-tabs.provider';
import { FilteredFieldContext } from '../../../../providers/filtered-field.provider';
import './CanvasFormHeader.scss';
import { FormTabsModes } from './canvasformtabs.modes';

interface CanvasFormHeaderProps {
  nodeId: string;
  title?: string;
  nodeIcon?: string;
  onClose?: () => void;
}

export const CanvasFormHeader: FunctionComponent<CanvasFormHeaderProps> = (props) => {
  const { filteredFieldText, onFilterChange } = useContext(FilteredFieldContext);
  const canvasFormTabsContext = useContext(CanvasFormTabsContext);

  return (
    <>
      <Grid hasGutter>
        <GridItem className="form-header" span={11}>
          <img className={`form-header__icon-${props.nodeId}`} src={props.nodeIcon} alt="icon" />
          <Title className="form-header__title" headingLevel="h2">
            {props.title}
          </Title>
        </GridItem>
        <GridItem span={1}>
          <Button data-testid="close-side-bar" variant="plain" icon={<TimesIcon />} onClick={props.onClose} />
        </GridItem>
      </Grid>

      {canvasFormTabsContext && (
        <ToggleGroup aria-label="Single selectable form tabs" className="form-tabs">
          {Object.entries(FormTabsModes).map(([mode, tooltip]) => (
            <ToggleGroupItem
              title={tooltip}
              key={mode}
              text={mode}
              buttonId={mode}
              isSelected={canvasFormTabsContext.selectedTab === mode}
              onChange={canvasFormTabsContext.onTabChange}
            />
          ))}
        </ToggleGroup>
      )}

      <SearchInput
        className="filter-fields"
        placeholder="Find properties by name"
        data-testid="filter-fields"
        value={filteredFieldText}
        onChange={onFilterChange}
        onClear={onFilterChange}
      />
    </>
  );
};
