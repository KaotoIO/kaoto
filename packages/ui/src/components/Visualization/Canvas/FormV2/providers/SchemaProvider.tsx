import { createContext, FunctionComponent, PropsWithChildren, useContext, useMemo } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';
import { resolveSchemaWithRef } from '../../../../../utils';
import { SchemaDefinitionsContext } from './SchemaDefinitionsProvider';

export interface SchemaContextValue {
  schema: KaotoSchemaDefinition['schema'];
  definitions: Record<string, KaotoSchemaDefinition['schema']>;
}

export const SchemaContext = createContext<SchemaContextValue>({ schema: {}, definitions: {} });

export const SchemaProvider: FunctionComponent<PropsWithChildren<{ schema: KaotoSchemaDefinition['schema'] }>> = ({
  schema,
  children,
}) => {
  const { definitions, omitFields } = useContext(SchemaDefinitionsContext);

  const value = useMemo(() => {
    const resolvedSchema = resolveSchemaWithRef(schema, definitions);

    if (Array.isArray(resolvedSchema.anyOf)) {
      resolvedSchema.anyOf = resolvedSchema.anyOf.map((anyOfSchema) => resolveSchemaWithRef(anyOfSchema, definitions));
    }
    if (Array.isArray(resolvedSchema.oneOf)) {
      resolvedSchema.oneOf = resolvedSchema.oneOf.map((oneOfSchema) => resolveSchemaWithRef(oneOfSchema, definitions));
    }

    omitFields.forEach((field) => {
      delete resolvedSchema.properties?.[field];
    });

    return { schema: resolvedSchema, definitions };
  }, [definitions, omitFields, schema]);

  return <SchemaContext.Provider value={value}>{children}</SchemaContext.Provider>;
};
