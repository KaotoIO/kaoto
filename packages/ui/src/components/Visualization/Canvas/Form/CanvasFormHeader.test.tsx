import { render } from '@testing-library/react';
import { CanvasFormTabsProvider } from '../../../../providers';
import { CanvasFormHeader } from './CanvasFormHeader';

describe('CanvasFormHeader', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <CanvasFormTabsProvider>
        <CanvasFormHeader nodeId="nodeId" nodeIcon="nodeIcon" title="title" />
      </CanvasFormTabsProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
