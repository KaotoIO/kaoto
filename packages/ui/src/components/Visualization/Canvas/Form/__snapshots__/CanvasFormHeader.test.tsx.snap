// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanvasFormHeader renders correctly 1`] = `
<DocumentFragment>
  <div
    class="pf-v5-l-grid pf-m-gutter"
  >
    <div
      class="pf-v5-l-grid__item pf-m-11-col form-header"
    >
      <img
        alt="icon"
        class="form-header__icon-nodeId"
        src="nodeIcon"
      />
      <h2
        class="pf-v5-c-title pf-m-xl form-header__title"
        data-ouia-component-id="OUIA-Generated-Title-1"
        data-ouia-component-type="PF5/Title"
        data-ouia-safe="true"
      >
        title
      </h2>
    </div>
    <div
      class="pf-v5-l-grid__item pf-m-1-col"
    >
      <button
        aria-disabled="false"
        class="pf-v5-c-button pf-m-plain"
        data-ouia-component-id="OUIA-Generated-Button-plain-1"
        data-ouia-component-type="PF5/Button"
        data-ouia-safe="true"
        data-testid="close-side-bar"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="pf-v5-svg"
          fill="currentColor"
          height="1em"
          role="img"
          viewBox="0 0 352 512"
          width="1em"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    aria-label="Single selectable form tabs"
    class="pf-v5-c-toggle-group form-tabs"
    role="group"
  >
    <div
      class="pf-v5-c-toggle-group__item"
      title="Shows Required fields only"
    >
      <button
        aria-pressed="true"
        class="pf-v5-c-toggle-group__button pf-m-selected"
        id="Required"
        type="button"
      >
        <span
          class="pf-v5-c-toggle-group__text"
        >
          Required
        </span>
      </button>
    </div>
    <div
      class="pf-v5-c-toggle-group__item"
      title="Shows All fields"
    >
      <button
        aria-pressed="false"
        class="pf-v5-c-toggle-group__button"
        id="All"
        type="button"
      >
        <span
          class="pf-v5-c-toggle-group__text"
        >
          All
        </span>
      </button>
    </div>
    <div
      class="pf-v5-c-toggle-group__item"
      title="Shows Modified fields only"
    >
      <button
        aria-pressed="false"
        class="pf-v5-c-toggle-group__button"
        id="Modified"
        type="button"
      >
        <span
          class="pf-v5-c-toggle-group__text"
        >
          Modified
        </span>
      </button>
    </div>
  </div>
  <div
    class="pf-v5-c-text-input-group filter-fields"
    data-testid="filter-fields"
  >
    <div
      class="pf-v5-c-text-input-group__main pf-m-icon"
    >
      <span
        class="pf-v5-c-text-input-group__text"
      >
        <span
          class="pf-v5-c-text-input-group__icon"
        >
          <svg
            aria-hidden="true"
            class="pf-v5-svg"
            fill="currentColor"
            height="1em"
            role="img"
            viewBox="0 0 512 512"
            width="1em"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            />
          </svg>
        </span>
        <input
          aria-label="Search input"
          class="pf-v5-c-text-input-group__text-input"
          placeholder="Find properties by name"
          type="text"
          value=""
        />
      </span>
    </div>
  </div>
</DocumentFragment>
`;
